### PR TITLE
add DragonEggFormEvent (fixes #5110)

### LINKED ISSUE #5110: Programmable dragon egg block spawning
**Is your feature request related to a problem? Please describe.**
There is little control plugins (and server owners) have over whether the dragon egg should or shouldn't appear when the ender dragon is defeated.

**Describe the solution you'd like**
Addition of a `DragonEggAppearEvent`. (Better name ideas are welcome.) Whether that event is canceled or not determines whether the egg should appear or not. The event should have access to the `DragonBattle` instance, or, if that's not possible for whatever reason, at least the `previouslyKilled` flag.

**Describe alternatives you've considered**
 - Plugins might be able to listen to dragon's death using already existing events, somehow get the trophy position (it's not currently exposed in the API, but getting it is probably possible nevertheless), and set the block at that position manually. But this process is sub ideal at best.
 - Percentage based control through configuration files: the dragon egg should have a configurable chance to appear. (This is the original request of `@maverick#4689`)

If feedback is positive, I plan on making and PRing the necessary changes.

### PR DESCRIPTION
**Pull request by Trigary**
Has been tested, ready for review.

I've decided to extend `BlockFormEvent`, but this new event has its own `HandlerList`. I believe this the best way to go about it, but feel free to change my mind.

### CODE REVIEW

**Review by MiniDigger — APPROVED** [review]
lgtm

File: Spigot-Server-Patches/0666-add-DragonEggFormEvent.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
++import io.papermc.paper.event.block.DragonEggFormEvent; // Paper - DragonEggFormEvent
+ 
+ public class EnderDragonBattle {
+ 
+@@ -358,9 +359,24 @@ public class EnderDragonBattle {
+             this.bossBattle.setVisible(false);
+             this.generateExitPortal(true);
+             this.n();
++            // Paper start - DragonEggFormEvent
++            BlockPosition eggPosition = this.world.getHighestBlockYAt(HeightMap.Type.MOTION_BLOCKING, WorldGenEndTrophy.a);
```
Comment (Proximyst) [inline Spigot-Server-Patches/0666-add-DragonEggFormEvent.patch]: obfhelper

**Review by Proximyst — APPROVED** [review]
LGTM; poke me sometime to have it merged.

File: Spigot-Server-Patches/0666-add-DragonEggFormEvent.patch (hunk 1 above)
Comment (MiniDigger) [reply Spigot-Server-Patches/0666-add-DragonEggFormEvent.patch]: saw that too, but since its just copy pasted from below I didnt care

Comment (Trigary) [reply Spigot-Server-Patches/0666-add-DragonEggFormEvent.patch]: Well I added a obfhelper

**Pull request merged**